core/trino-main/src/main/java/io/trino/sql/ir/GenericLiteral.java

```diff
@@ -16,26 +16,82 @@
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Primitives;
+import com.google.errorprone.annotations.DoNotCall;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.type.DecimalType;
 import io.trino.spi.type.Type;
 
 import java.util.List;
 import java.util.Objects;
 
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.spi.type.TypeUtils.readNativeValue;
+import static io.trino.spi.type.TypeUtils.writeNativeValue;
 import static java.util.Objects.requireNonNull;
 
 public final class GenericLiteral
         extends Literal
 {
     private final Type type;
-    private final String value;
+    private final String stringValue;
+    private final Object rawValue;
+
+    public static GenericLiteral constant(Type type, Object rawValue)
+    {
+        checkArgument(
+                Primitives.wrap(type.getJavaType()).isAssignableFrom(rawValue.getClass()),
+                "Improper Java type (%s) for type '%s'",
+                rawValue.getClass().getName(),
+                type.toString());
+
+        return new GenericLiteral(type, null, rawValue);
+    }
 
     @JsonCreator
-    public GenericLiteral(Type type, String value)
+    @DoNotCall // For JSON deserialization only
+    public static GenericLiteral fromJson(
+            @JsonProperty Type type,
+            @JsonProperty Block rawValueAsBlock,
+            @JsonProperty String stringValue)
+    {
+        return new GenericLiteral(type, stringValue, readNativeValue(type, rawValueAsBlock, 0));
+    }
+
+    @Deprecated
+    public GenericLiteral(Type type, String stringValue)
+    {
+        this(verifyLegacyType(type), stringValue, null);
+    }
+
+    private static Type verifyLegacyType(Type type)
+    {
+        if (type.equals(TINYINT) ||
+                type.equals(SMALLINT) ||
+                type.equals(INTEGER) ||
+                type.equals(BIGINT) ||
+                type.equals(BOOLEAN) ||
+                type.equals(DOUBLE) ||
+                type instanceof DecimalType) {
+            throw new IllegalArgumentException("Call constant(%s, ...)".formatted(type));
+        }
+
+        return type;
+    }
+
+    public GenericLiteral(Type type, String stringValue, Object rawValue)
     {
         requireNonNull(type, "type is null");
-        requireNonNull(value, "value is null");
         this.type = type;
-        this.value = value;
+        this.stringValue = stringValue;
+        this.rawValue = rawValue;
```

Comment:
I'm probably going to use it for typed nulls in the future, too. Regardless, for now they need to be able to be null since I'm using the structure as a union. For legacy types that haven't been migrated to the new style of generic literals (i.e., with the string representation instead of the native carrier value), the old behavior remains.

I'll remove the stringValue field once I finish migrating the rest of the types.

Reply:
There's already a validation for the type of rawValue vs the expected carrier type in the method above:

```java
    public static GenericLiteral constant(Type type, Object rawValue)
    {
        checkArgument(
                Primitives.wrap(type.getJavaType()).isAssignableFrom(rawValue.getClass()),
                "Improper Java type (%s) for type '%s'",
                rawValue.getClass().getName(),
                type.toString());

        return new GenericLiteral(type, null, rawValue);
    }
```